cmd/activator/main.go

```diff
@@ -133,6 +150,84 @@ func main() {
 		logger.Fatalw("Failed to create stats reporter", zap.Error(err))
 	}
 
+	// Set up signals so we handle the first shutdown signal gracefully.
+	stopCh := signals.SetupSignalHandler()
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, defaultResyncInterval)
+	servingInformerFactory := servinginformers.NewSharedInformerFactory(servingClient, defaultResyncInterval)
+	endpointInformer := kubeInformerFactory.Core().V1().Endpoints()
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
+	go revisionInformer.Informer().Run(stopCh)
+
+	params := queue.BreakerParams{QueueDepth: breakerQueueDepth, MaxConcurrency: breakerMaxConcurrency, InitialCapacity: 0}
+
+	// Return the number of endpoints, 0 if no endpoints are found.
+	endpointsGetter := func(revID activator.RevisionID) (int32, error) {
+		endpointKey := cache.ExplicitKey(revID.Namespace + "/" + reconciler.GetServingK8SServiceNameForObj(revID.Name))
+		endpoints, exists, err := endpointInformer.Informer().GetIndexer().Get(endpointKey)
+		if err != nil {
+			return 0, fmt.Errorf("Error getting endpoints for revision %v", revID)
+		}
+		if !exists {
+			// No endpoints exist yet.
+			return 0, nil
+		}
+		if exists && endpoints != nil {
+			endpoint := endpoints.(*corev1.Endpoints).Subsets
+			addresses := activator.EndpointsAddressCount(endpoint)
+			return int32(addresses), nil
+		}
+		return 0, nil
+	}
+
+	// Return the revision from the observer.
+	// Additionally to observer errors an error is returned if either no revision exists yet
+	// or we could not cast the interface to Revision for some reason.
+	revisionGetter := func(revID activator.RevisionID) (revision *v1alpha12.Revision, err error) {
+		revisionKey := cache.ExplicitKey(revID.Namespace + "/" + revID.Name)
+		rev, ok, err := revisionInformer.Informer().GetIndexer().Get(revisionKey)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			return nil, fmt.Errorf("Error getting revision for %v", revID)
+		}
+		if revision, ok = rev.(*v1alpha12.Revision); !ok {
+			return nil, fmt.Errorf("Error casting revision %v", revID)
+		}
+		return revision, err
+	}
+
+	throttlerParams := activator.ThrottlerParams{BreakerParams: params, Logger: logger, GetEndpoints: endpointsGetter, GetRevision: revisionGetter}
+	throttler := activator.NewThrottler(throttlerParams)
+
+	// Update/create the breaker in the throttler when the number of endpoints changes.
+	endpointInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Do we need to listen for "Add" here as well?

Reply:
As far as I remember the endpoint is getting added with 0 value and then it is updated to 1. So we don't need to watch for the add.

Reply:
I'd say let's be better safe then sorry. We have utility functions to transform an `Add` handler into an `Update` handler.

In particular, your UpdateHandler already ignores the first argument we can straight ahead drop it from the signature and change this to this:

```golang
endpointInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
	AddFunc:    activator.UpdateEndpoints(throttler),
	UpdateFunc: controller.PassNew(activator.UpdateEndpoints(throttler)),
})
```

WDYT?

Reply:
Makes sense, was not aware of the PassNew trick.